src/hotspot/share/opto/ifnode.cpp

```diff
@@ -838,7 +838,8 @@ bool IfNode::has_only_uncommon_traps(ProjNode* proj, ProjNode*& success, ProjNod
       ciMethod* dom_method = dom_unc->jvms()->method();
       int dom_bci = dom_unc->jvms()->bci();
       if (!igvn->C->too_many_traps(dom_method, dom_bci, Deoptimization::Reason_unstable_fused_if) &&
-          !igvn->C->too_many_traps(dom_method, dom_bci, Deoptimization::Reason_range_check)) {
+          !igvn->C->too_many_traps(dom_method, dom_bci, Deoptimization::Reason_range_check) &&
+          igvn->C->remove_unstable_if_trap(dom_unc)) {
```

Comment:
This should be moved to `IfNode::merge_uncommon_traps`.

Reply:
I tried that in the first place. It turns out that situation is complex. The following 3 functions take place in order. 
has_only_uncommon_traps() => fold_compares_helper() => merge_uncommon_traps()

`IfNode::has_only_uncommon_traps` is the last stop before modifying code. if we moved the predicate "remove_unstable_if_trap(dom_unc)" to `merge_uncommon_traps()`, it would be too late when it returns true. We have to rollback code change in `fold_compares_helper`.